packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -2549,21 +2549,29 @@ class RenderPointerListener extends RenderProxyBoxWithHitTestBehavior {
   MouseTrackerAnnotation get hoverAnnotation => _hoverAnnotation;
 
   void _updateAnnotations() {
-    if (_hoverAnnotation != null && attached) {
-      RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation);
-    }
+    MouseTrackerAnnotation newAnnotation;
     if (_onPointerEnter != null || _onPointerHover != null || _onPointerExit != null) {
-      _hoverAnnotation = MouseTrackerAnnotation(
+      newAnnotation = MouseTrackerAnnotation(
         onEnter: _onPointerEnter,
         onHover: _onPointerHover,
         onExit: _onPointerExit,
       );
-      if (attached) {
-        RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation);
-      }
-    } else {
-      _hoverAnnotation = null;
     }
+    if (newAnnotation == _hoverAnnotation) {
+      return;
+    }
+    if (_hoverAnnotation != null && attached) {
+      // Detach the existing annotation.
+      RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation);
+    }
+    _hoverAnnotation = newAnnotation;
+    if (_hoverAnnotation != null && attached) {
+      // Attach the new annotation.
+      RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation);
+    }
```

Comment:
L2560-2571 took me a while to understand, I think a clearer way is
```dart
if (newAnnotation == null && _hoverAnnotation == null)
  return
if (attached) {
  if (_hoverAnnotation != null) {
    RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation);
  }
  if (newAnnotation != null) {
    RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation);
  }
}
_hoverAnnotation = newAnnotation;
```

Reply:
That is clearer, except that I'd rather that `_hoverAnnotation` be up to date by the time `attachAnnotation` is called.  I'll play with it some more and see if I can improve the readability.

Reply:
I basically reverted it to what it was before, since that was clearer.  I added an assert which makes sure that something changed, since that was part of what I was trying to address with the reorg of the logic, but in reality won't happen, since the callers all check and see that something changed before calling it.